src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs

```diff
@@ -362,7 +368,15 @@ public override object Invoke(ServiceProvider provider)
 
             public override Expression Build(Expression provider)
             {
-                return base.Build(Expression.Field(provider, "_root"));
+                // PERF: Singleton services once created will always be available in the lookup dictionary
+                // Hence it is cheaper to directly call its Invoke method that will do the lookup rather than 
+                // building an expression tree that does the same.
+                // This change reduces the memory allocations during the start up because, expressions accessing private members
+                // involves expensive code access security checks for Desktop CLR.
+                return Expression.Call(
+                    Expression.Constant(this, typeof(SingletonCallSite)),
+                    InvokeMethodInfo,
+                    provider);
```

Comment:
Is the memory allocation proportional to the number of expression nodes, or to the number of calls to lambda compile?

Assuming it's proportional to the number of calls to compile - I don't think you need to change this method at all if the top-level Build isn't called for direct calls to GetService of singletons.